lib/mongoid/criteria/queryable/storable.rb

```diff
@@ -48,7 +48,7 @@ def add_field_expression(field, value)
             if value.is_a?(Hash) && selector[field].is_a?(Hash) &&
               value.keys.all? { |key|
                 key_s = key.to_s
-                key_s[0] == ?$ && !selector[field].key?(key_s)
+                key_s[0] == ?$ && !selector[field].transform_keys(&:to_s).key?(key_s)
               }
```

Comment:
We could code-golf this, but this whole thing could be made cleaner and avoid transforming all the selector[field] keys for each value key

```suggestion
              selector_keys = selector[field].transform_keys(&:to_s)
              value.transform_keys(&:to_s).keys.all? { |k| k[0] == ?$ && !selector_keys.key?(k) }
```